### DISCUSSION TITLE
How to set position use anchor point Vec2(0.5, 0.5) like cocos creator?

### DISCUSSION THREAD

**Discussion by antn9x** [axmolengine/axmol#998] [General]
<img width="483" alt="image" src="https://user-images.githubusercontent.com/2052604/209814662-b7d6f02e-f878-4338-9912-3d6fd84ab24e.png">
How to set position in center of parent node with anchor point (0.5, 0.5) like cocos creator do?

**Comment by DelinWorks** [comment]
you can use `node->setAnchor(0.5F, 0.5F);`

**Reply by antn9x** [reply]
It doesn't work, still draw child not in BOTTOM LEFT position.
<img width="217" alt="image" src="https://user-images.githubusercontent.com/2052604/209862292-387afc96-68f3-48ed-af2a-5e2ed195c729.png">

**Reply by aismann** [reply]
Can we get your source snippet?

**Reply by antn9x** [reply]
<img width="782" alt="image" src="https://user-images.githubusercontent.com/2052604/209862806-2a1839ed-d09c-452a-a9a8-39aa5c6b0d50.png">
I setAnchorPoint(Vec2(0.5,0.5)); in CCNode to apply all new nodes. And set Hello world as child of sprite in Helloworld project.

**Reply by aismann** [reply]
Why you set it not on each new Sprite?
If you change it on ALL nodes also the parent node ist changed ( Not tested)

**Reply by antn9x** [reply]
Because I want to have behavior like cocos creator with default is anchor (0.5, 0.5) and child node will draw at center of parent node.

**Reply by DelinWorks** [reply]
you don't have to change the `node.cpp` file to anchor a node. You should make that node anchor to 0.5 from within your project code not the engine's, not its source, everything derives from node so if you set anchor to 0.5 in `node.cpp` then everything including the camera will be of 0.5 anchor so there is no effect, you should make that change in the helloworld class of the project.

**Reply by antn9x** [reply]
Do you have any sample code worked?

**Reply by DelinWorks** [reply]
you need to locate `HelloWorldScene.cpp` and change the code where it creates the hello world sprite so that it anchors to 0.5, at first you will notice that it's indeed in the middle but that's because the position was shifted beforehand, you need to comment the set position line and set the anchor to 0.5 like this:

```
    // add "HelloWorld" splash screen"
    auto sprite = Sprite::create("HelloWorld.png"sv);
    if (sprite == nullptr)
    {
        problemLoading("'HelloWorld.png'");
    }
    else
    {
        // position the sprite on the center of the screen
        // this should be commented to test anchoring 
        // sprite->setPosition(Vec2(visibleSize.width / 2 + origin.x, visibleSize.height / 2 + origin.y));
        sprite->setAnchorPoint(Vec2(0.5,0.5));

        // add the sprite as a child to this layer
        this->addChild(sprite, 0);

        auto drawNode = DrawNode::create();
        drawNode->setPosition(Vec2(0, 0));

        addChild(drawNode);

        drawNode->drawRect(safeArea.origin, safeArea.origin + safeArea.size, Color4F::BLUE);
    }
```

**Comment by antn9x** [comment]
That's not what I need. I want add label to sprite, and then label in center of sprite without set position.

**Reply by DelinWorks** [reply]
Okay I'll see what I can do

**Reply by DelinWorks** [reply]
It seems like `setAnchorPoint` moves the point of the node and doesn't shift the sprite itself that's why attaching a label to a sprite always makes its position in the bottom left of the sprite, the sprite class should have a separate function like `setAnchorPointSprite` and that function should shift the sprite, not the node that holds the sprite. I'll investigate more

**Reply by antn9x** [reply]
I think you was right, now we need to check how to render sprite by anchor point.

**Reply by DelinWorks** [reply]
I found a fix! go to the engine's code in `CCSprite.cpp` and make this change to the function `setVertexCoords`:

![image](https://user-images.githubusercontent.com/45469625/209939280-2627a50d-8ee7-4a63-8f3a-d0e457e5bed9.png)

This should be the right way to shift a sprite to the center, not with `setAnchorPoint`
You can see now that the label appears in the center of the sprite:

![image](https://user-images.githubusercontent.com/45469625/209939403-f78f3470-7457-4fcf-af4f-b83ad83b21cb.png)
![image](https://user-images.githubusercontent.com/45469625/209939927-65974b2a-6639-4b2b-94d6-e6a6385371e8.png)

of course, the letters in the label are misplaced because we modified the engine's code directly and the code applies to everything renderable so I think a pull request that adds a separate function should fix this.

**Reply by antn9x** [reply]
I think this is correct answer for my question, thank you so much. We can improve this code to right section, hard code here is not good enough.

**Reply by DelinWorks** [reply]
We'll let @rh101 decide what's the right fix for this problem, do you think a `setAnchorPointSprite` function is any good?

**Reply by rh101** [reply]
I just read this entire thread. First thing, this engine isn't based on Cocos Creator, but rather on Cocos2d-x v4, so we can't expect the same behaviour of nodes as Cocos Creator.  If something isn't working the same as Cocos Creator, it doesn't mean it's a bug in Axmol.

> We'll let @rh101 decide what's the right fix for this problem, do you think a `setAnchorPointSprite` function is any good?

I'm not certain why you think it's my decision for this, but my opinion regarding it is that we shouldn't be introducing custom methods like this.  We can add any kind of child node to a another node, so having `setAnchorPointSprite` doesn't make sense, since it can be any node, and not just a sprite.

Now, regarding this specific issue, the anchor point is not the same as the position.  The default position is, I assume, set to 0,0, for any newly created node, and when this is added as a child to another node, it'll be drawn at 0,0, and adjusted based on the anchor point of that child object.  You can't get around not setting the position and whatever other methods need to be called.

**Reply by DelinWorks** [reply]
it's actually a flaw in the engine, instead of moving the vertices of a quad so that it's in the center, the node itself is used to anchor the sprite which is completely wrong and doesn't make sense. it's actually a thing I wanted to point out for so long and because sometimes it becomes infuriating when building UI systems with the current sprite anchoring system, what about making the function `setAnchorPoint` override the node class and position the sprite vertices accordingly? of course, that would break compatibility but there should be another solution, I asked for your decision/opinion on the function naming and future-proofing the fix, not on whether this fix should exist or not. my bad I didn't clarify.

**Reply by rh101** [reply]
What you're suggesting is that the 0,0 position of child nodes is in the center of a parent node. Is my understanding of that correct? 

What it would do is affect a lot more than just the visual position of the child node. Have you tested the impact it would have on the world position of that child node, and also the target position for touch handling etc.? 

> when building UI systems with the current sprite anchoring system, what about making the function setAnchorPoint override the node class and position the sprite vertices accordingly?

If sprites are to be used as UI elements with different behavior of the anchor point, then `Sprite` should sub-classed and a UI system built off that.  

Unless I'm not quite understanding what the issue is here, in my opinion, how they work at the moment is correct, and a position of 0,0 on a child sprite should be drawn at the bottom left of the parent node; the child sprite with 0,0 position should **not** be drawn in the middle of the parent.

**Reply by DelinWorks** [reply]
I agree with you but for some use cases it's better to have an option in a method that makes you choose what type of anchor you want

**Comment by aismann** [comment]
Have a look at
cpp-test:
24: Testing Sprite
25. Testing SpriteBatchNode

**Reply by antn9x** [reply]
I read example code, but not what i need to sprite render by anchor point.

**Reply by rh101** [reply]
> I read example code, but not what i need to sprite render by anchor point.

The behaviour of Cocos2d-x and Axmol is not the same as Cocos Creator. This is not an issue in this engine, and you can't avoid not setting the position.

**Comment by aismann** [comment]
Try this (answer on cocos2dx forum):
```
auto parentNode = Node::create();
parentNode->setContentSize(Size(200, 200));

// child
auto sprite = Sprite::create("sample.png");
sprite->setAnchorPoint(Vec2::ANCHOR_MIDDLE);  // (0.5, 0.5)
sprite->setPosition(parentNode->getContentSize() / 2);

parentNode->addChild(sprite);
```

**Comment by aismann** [comment]
Here is a working code (setPosition is needed):
```
        auto sprite = Sprite::create("Images/grossini_dance_08.png");
        sprite->setPosition(Vec2(s.width / 2, s.height / 2));
        addChild(sprite);
        auto label = Label::createWithTTF("Hello World", "fonts/Marker Felt.ttf", 24);
        label->setPosition(sprite->getPosition());
        addChild(label);
```
Using an  own method can hidden this. 
![image](https://user-images.githubusercontent.com/8652787/210004159-a39bf812-7ce5-4088-999a-04faaadb2f6e.png)

**Reply by antn9x** [reply]
It should be `label->setPosition(sprite-> getContentSize() / 2);`

**Reply by antn9x** [reply]
I did this but it need to use `setPosition` that is not what i want.

**Reply by rh101** [reply]
> It should be `label->setPosition(sprite-> getContentSize() / 2);`

No, that's not correct for the snippet of code you're referring to.  If you read the code, you will notice that label is not a child of the sprite, but at the same level as the sprite (so both are children of the same parent).  That would mean the original code posted is correct, being:
`label->setPosition(sprite->getPosition());` (assuming the anchor point of the label is Vec(0.5, 0.5)).

For the first snippet posted by aismann, you'll notice that the label is a child of the sprite, so it correctly has this:
`sprite->setPosition(parentNode->getContentSize() / 2);`

So, both snippets of code are correct for what they're trying to achieve.

> I did this but it need to use `setPosition` that is not what i want.

What you want and what is correct seem to be clashing here.  There is absolutely no bug with the current implementation in this engine. So, once again, you cannot avoid `setPosition`.

If you want it to behave in the way you keep insisting on, then implement a sub-class of `Label`, and override all the methods you need to get the functionality you desire.
